src/connector/src/sink/kinesis.rs

```diff
@@ -191,39 +189,176 @@ impl KinesisSinkWriter {
+                let total_count = self.entries.len();
+                let mut start_idx = 0;
+
+                let mut throttle_delay = None;
+
+                while start_idx < total_count {
+                    // 1. Prepare the records to be sent
+
+                    // The maximum possible number of records that can be sent in this iteration.
+                    // Can be smaller than this number of the total payload size exceeds `MAX_TOTAL_RECORD_PAYLOAD_SIZE`
+                    let max_record_count = min(MAX_RECORD_COUNT, total_count - start_idx);
+                    let mut records = Vec::with_capacity(max_record_count);
+                    let mut total_payload_size = 0;
+                    for i in start_idx..(start_idx + max_record_count) {
+                        let (record, size) = &self.entries[i];
+                        if *size >= MAX_SINGLE_RECORD_PAYLOAD_SIZE {
+                            warn!(
+                                size,
+                                partition = record.partition_key,
+                                "encounter a large single record"
+                            );
+                        }
+                        if total_payload_size + *size < MAX_TOTAL_RECORD_PAYLOAD_SIZE {
+                            total_payload_size += *size;
+                            records.push(record.clone());
+                        } else {
+                            break;
+                        }
+                    }
+                    if records.is_empty() {
+                        // at least include one record even if its size exceed `MAX_TOTAL_RECORD_PAYLOAD_SIZE`
+                        records.push(self.entries[start_idx].0.clone());
+                    }
+
+                    // 2. send the records and handle the result
+                    let record_count = records.len();
+                    match self
+                        .client
+                        .put_records()
+                        .stream_name(&self.stream_name)
+                        .set_records(Some(records))
+                        .send()
+                        .await
+                    {
+                        Ok(output) => {
+                            if record_count != output.records.len() {
+                                return Err(SinkError::Kinesis(anyhow!("request record count {} not match the response record count {}", record_count, output.records.len())));
+                            }
+                            // From the doc of `put_records`:
+                            // A single record failure does not stop the processing of subsequent records. As a result,
+                            // PutRecords doesn’t guarantee the ordering of records. If you need to read records in the same
+                            // order they are written to the stream, use PutRecord instead of PutRecords, and write to the same shard.
```

Comment:
Shall we acknowledge this limitation in doc as there is currently no user option to opt for ordering rather than batching?